### DISCUSSION TITLE
Read a table with 300,000,000 million records

### DISCUSSION THREAD

**Discussion by patriciomh1965** [serenity-is/Serenity#6095] [Help]
Hello community

I need to read a table that has a large number of records, the page is timeout. Is there a criterion or a solution to read the table?

I attach an image with the information that it should show, but as I said it falls out of time.
![Demo](https://user-images.githubusercontent.com/42218266/142217839-33c97359-0596-4e08-b902-ef0207b231b2.png)

I appreciate any help

Atte.

**Comment by edwardch** [comment]
There are some things you can do to help speed it up, but the problem is most likely more on the sql/database side of things.

From a project perspective:

- Exclude row count
- Don't allow user to sort on certain fields that might slow the query `[Sortable(false)]`
- Remove unnecessary joins if possible
- Enforce the user to apply multiple/any grid filters (to narrow down data, see default filter values in demo)
- Consider converting the List method to use a stored procedure, which can be optimized more easily
- etc

From a database perspective:

- Inspect queries (SQL profiler is your friend), and apply indexes if necessary
- Ensure data types are as small as possible
- Consider partitioning your table
- Get a bigger server :)

I have a table with about 250M records and these points certainly helped.

Also see here for reference: https://github.com/serenity-is/Serenity/issues/4399

HTH

**Reply by patriciomh1965** [reply]
Hello

Thanks for your help, I wanted to see if you have programmed a call to a stored procedure that receives parameters. I have tried but so far it has not worked for me.

I would appreciate your help

**Reply by edwardch** [reply]
I tried looking for the stored procedure code from the demo but can't seem to find it right now, @VictorTomaili is the demo running different code from common-features?

@patriciomh1965 you can set it up something like this:

Create your stored procedure, and make sure the returned columns match your columns in YourRow.cs.

Then create a custom list request for it like so:
```cs
public class YourCustomListRequest : ListRequest
{
    public int? ClientId { get; set; }
    public int? OtherParameter1 { get; set; }
    public int? OtherParameter2 { get; set; }
}
```

In YourRepository.cs update the List method
```cs
public ListResponse<MyRow> List(IDbConnection connection, YourCustomListRequest request)
{
    var data = connection.Query<MyRow>("[dbo].[YourStoredProcedure]",
        param: new
        {
            request.ClientId,
            request.OtherParameter1,
            request.OtherParameter2
        },
        commandType: CommandType.StoredProcedure);

    var response = new ListResponse<MyRow>
    {
        Entities = (List<MyRow>)data
    };
    return response;
}
```

And then on YourGrid.ts, override the quick filters:
```ts
protected getQuickFilters() {
    var filters = super.getQuickFilters();

    filters.push({
        type: Serenity.LookupEditor,
        options: {
            lookupKey: YourProject.Clients.ClientRow.lookupKey
        },
        field: 'ClientId',
        title: 'Client',
        handler: w => {
            (this.view.params as YourCustomListRequest).ClientId = Q.toId(w.value);
        }
    });

    // Push filters for OtherParameter1 and OtherParameter2
    // Look around the demo code for examples of different data types
    return filters;
}
```

**Reply by edwardch** [reply]
@patriciomh1965 if this helps, please consider adding a wiki entry with steps :)

**Reply by VictorTomaili** [reply]
@edwardch demo running with StartSharp

**Reply by edwardch** [reply]
Ahh right, that makes sense. Thanks Victor

**Comment by patriciomh1965** [comment]
Thanks edwardch

I am going to implement the solution, if I have doubts I will turn you back to rude.
